### PR TITLE
ci: slightly rework the docs-check workflow.

### PR DESCRIPTION
**Pull request by obnoxxx**
**Descrition of changes:** 

This reworks the docs-check ci workflow in several ways:

* It  renames the make target `check-docs` to the smore systematic `check.docs`.
* It simplifies the modified files validation script, adds a `docs` mode, and uses the script in `make check.docs`.

Overall, the workflow and local make targets are more systematic and consistent with this change.

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [x] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

### CODE REVIEW

File: tests/scripts/validate_modified_files.sh [hunk 1]
```diff
@@ -9,26 +9,31 @@ MOD_ERR="changes found by mod.check. You may need to run make clean"
 CRD_ERR="changes found by 'make crds'. please run 'make crds' locally and update your PR"
 BUILD_ERR="changes found by make build', please commit your go.sum or other changed files"
 HELM_ERR="changes found by 'make gen-rbac'. please run 'make gen-rbac' locally and update your PR"
+DOCS_ERR="changes found by 'make docs'. please run 'make docs'locally and update your PR"
```
Comment (sp98) [inline tests/scripts/validate_modified_files.sh:12]:
```suggestion
DOCS_ERR="changes found by 'make docs'. please run 'make docs' locally and update your PR"
```

File: tests/scripts/validate_modified_files.sh [hunk 2]
```diff
@@ -9,26 +9,31 @@ MOD_ERR="changes found by mod.check. You may need to run make clean"
...
 #############
 function validate(){
-  git=$(git status --porcelain)
-  for file in $git; do
-    if [ -n "$file" ]; then
-      echo "$1"
-      echo "$git"
-      git diff
+  ERR=$1
+  git diff  --exit-code
```
Comment (sp98) [inline tests/scripts/validate_modified_files.sh:20]:
```suggestion
  git diff --exit-code
```

File: tests/scripts/validate_modified_files.sh [hunk 3]
```diff
@@ -9,26 +9,31 @@ MOD_ERR="changes found by mod.check. You may need to run make clean"
...
+  ret=$?
+  if [ "$ret" != "0" ]; then
+    echo "modidied files detected."
```
Comment (sp98) [inline tests/scripts/validate_modified_files.sh:23]:
```suggestion
    echo "modified files detected."
```

**Review by sp98 — CHANGES_REQUESTED** [review]

File: tests/scripts/validate_modified_files.sh [hunk 4]
```diff
@@ -9,26 +9,31 @@ MOD_ERR="changes found by mod.check. You may need to run make clean"
...
 #############
 function validate(){
-  git=$(git status --porcelain)
-  for file in $git; do
-    if [ -n "$file" ]; then
-      echo "$1"
-      echo "$git"
-      git diff
+  ERR=$1
+  git diff --exit-code
```
Comment (sp98) [inline tests/scripts/validate_modified_files.sh:20]: (Nothing to fix but just an observation) Earlier, the function would only run `git status --porcelain` to get the changed files and then print those files. Now it runs `git diff --exit-code` as well as `git status --porcelain`. So we are making an extra git call in the refracted function.

File: tests/scripts/validate_modified_files.sh [hunk 5]
```diff
@@ -9,26 +9,31 @@ MOD_ERR="changes found by mod.check. You may need to run make clean"
...
-    fi
-  done
+  fi
 }
 
 ########
 # MAIN #
 ########
 case "$1" in
+docs)
```
Comment (sp98) [inline tests/scripts/validate_modified_files.sh:34]: should `echo $"Usage: $0 {codegen|modcheck|crd|build|gen-rbac}" `  reflect `docs` as well ?

Comment (obnoxxx) [reply tests/scripts/validate_modified_files.sh:34]: @sp98  wrote:

> should `echo $"Usage: $0 {codegen|modcheck|crd|build|gen-rbac}" ` reflect `docs` as well ?

good point, thanks! done.

File: tests/scripts/validate_modified_files.sh (hunk 4 above)
Comment (sp98) [reply tests/scripts/validate_modified_files.sh:20]: closing this.

File: .github/workflows/docs-check.yml [hunk 6]
```diff
@@ -45,7 +45,7 @@ jobs:
       - name: Check helm-docs
         run: make check-helm-docs
       - name: Check docs
-        run: make check-docs
+        run: make check.docs
```
Comment (sp98) [inline .github/workflows/docs-check.yml:48]: This command needs to be changed in the documentation as well - https://github.com/rook/rook/blob/master/Documentation/Contributing/documentation.md#making-docs

Comment (sp98) [inline .github/workflows/docs-check.yml:48]: All other changes look good. But I'm not too sure about the target naming convention of `make check.docs`.  So I'll wait for other's opinion on this as well.

Comment (obnoxxx) [reply .github/workflows/docs-check.yml:48]: @sp98  wrote:

> All other changes look good. But I'm not too sure about the target naming convention of `make check.docs`. So I'll wait for other's opinion on this as well.

I don't claim that `check.docs` is even  a convention. In rook, we have a few other target groups stuctured with dots, like `go.vet`, `go.fmt`, `go.mod.check, etc.

So I thought I'd introduce a new case of this pattern with  `check.foo`.

Comment (obnoxxx) [reply .github/workflows/docs-check.yml:48]: done

File: tests/scripts/validate_modified_files.sh [hunk 7]
```diff
@@ -9,26 +9,31 @@ MOD_ERR="changes found by mod.check. You may need to run make clean"
...
-  git=$(git status --porcelain)
-  for file in $git; do
-    if [ -n "$file" ]; then
-      echo "$1"
-      echo "$git"
-      git diff
+  ERR=$1
+  git diff --exit-code
+  ret=$?
+  if [ "$ret" != "0" ]; then
```
Comment (parth-gr) [inline tests/scripts/validate_modified_files.sh:22]: have you tested this,

Any easy would be updating the docs in the second commit and not running `make check.docs`

**Review by parth-gr — CHANGES_REQUESTED** [review]

File: Documentation/Contributing/documentation.md [hunk 8]
```diff
@@ -35,4 +35,4 @@ When previewing, now you can navigate your browser to [http://127.0.0.1:8000/](h
 ## Making docs
 
 [helm-docs](https://github.com/norwoodj/helm-docs) is a tool that generates the documentation for a helm chart automatically. If there are changes in the helm chart, the developer needs to run `make docs` (to run helm-docs) and check in the resulting autogenerated files.
-To make it easy to check locally for uncommitted changes generated by `make docs`, an additional `make` target exists: simply running `make check-docs` will run the docs auto-generation and will complain if this produces uncommitted changes to doc files. It is therefore a good habit to always run `make check-docs` locally before creating or updating a PR.
+To make it easy to check locally for uncommitted changes generated by `make docs`, an additional `make` target exists: simply running `make check.docs` will run the docs auto-generation and will complain if this produces uncommitted changes to doc files. It is therefore a good habit to always run `make check-docs` locally before creating or updating a PR.
```
[the remark this answers is not in the document]
Comment (obnoxxx) [reply Documentation/Contributing/documentation.md]: @parth-gr  wrote:
> > always run `make check-docs`
> 
> PLease update second instance on the same line


done!

**Review by sp98 — APPROVED** [review]

File: tests/scripts/validate_modified_files.sh (hunk 7 above)
Comment (parth-gr) [reply tests/scripts/validate_modified_files.sh:22]: if you have tested it then its fine,
I was just suggesting to add a second commit with some doc change and not running `make cmd`

**Review by parth-gr — APPROVED** [review]
lgtm

**Comment by parth-gr** [comment]
Please add the testing results done lcoally

File: tests/scripts/validate_modified_files.sh [hunk 9]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env -S bash
-set -ex
```
Comment (subhamkrai) [inline tests/scripts/validate_modified_files.sh:2]: why removing this?

**Review by subhamkrai — COMMENTED** [review]
One comment, I'm not sure if there discussion or issue for this. I'm just trying to understand the requirements

File: .github/workflows/docs-check.yml (hunk 6 above)
Comment (subhamkrai) [reply .github/workflows/docs-check.yml:48]: @obnoxxx I'm not able to find doc to support my case but in rook we also, `make docs.build` `make docs.preview` etc, so not sure which is right format for this

File: tests/scripts/validate_modified_files.sh (hunk 9 above)
Comment (obnoxxx) [reply tests/scripts/validate_modified_files.sh:2]: @subhamkrai  wrote:
> why removing this?

1. I prefer having explicit error exits accompanied by error messages  in scripts instead of implicit exits without helpful messages, due to `set -e`.
2. I find the tracing of `set -x` very confusing: it obfuscates the important messages the script prints and makes debungging very hard, IMHO.


short summay: `set -ex` makes script output extremely hard to understand for me and removing it is almost always the first thing I do to a script.  So it is mostly personal preference.

**Comment by obnoxxx** [comment]
@subhamkrai  wrote:
> One comment, I'm not sure if there discussion or issue for this. I'm just trying to understand the requirements

There is no discussion or issue.

Motivation:

This is the beginning of a series of changes to make the ci workflows and related  make targets more systematic.

I want a developer to be able to run checks from ci workflows locally before pr creation or updation, ideally with a systematically named make target that is invoked verbatim in the corresponding ci workflow.

Hope this helps.

@subhamkrai, do you want me to create a corresponding issue to link here?

**Comment by obnoxxx** [comment]
@parth-gr  wrote:
> Please add the testing results done lcoally

apart from running the new  make target locally, e. g. `make check.docs`, I ran a few steps to test  the changes to the script directly:

First: `./tests/scripts/validate_modified_files.sh` to see Usage output

Second: a few steps to trigger the error condition without  running the entire docs creation:

* modify `README.md` with the editor
* run the script: 
```console
$ ./tests/scripts/validate_modified_files.sh docs
modified files detected.
 M README.md
?? output.txt
?? tags
changes found by 'make docs'. please run 'make docs' locally and update your PR
```
Result: clean unclobbered output.

**Comment by obnoxxx** [comment]
The last update changed `git --exit-code`in the script to `git --quiet` to further clear the output.

**Comment by subhamkrai** [comment]
> @subhamkrai, do you want me to create a corresponding issue to link here?

@obnoxxx Yes, please open an issue and get everyone feedback

**Comment by obnoxxx** [comment]
@subhamkrai  wrote:
> > @subhamkrai, do you want me to create a corresponding issue to link here?
> 
> @obnoxxx Yes, please open an issue and get everyone feedback

I created a first, overarching issue for this and subsequent changes:

https://github.com/rook/rook/issues/14686

This PR addresses the first item listed there, while #14674  addresses the second.

File: tests/scripts/validate_modified_files.sh (hunk 7 above)
Comment (BlaineEXE) [reply tests/scripts/validate_modified_files.sh:22]: Following the conversation around `set -ex`, let's not change this. The script is working today, and from a code/project maintenance perspective, this seems like unnecessary risk of breaking corner cases when running the script. We already have enough CI instability, and I don't want to risk more over scripting preferences.

Comment (obnoxxx) [reply tests/scripts/validate_modified_files.sh]: @BlaineEXE  wrote:
> Following the conversation around `set -ex`, let's not change this. The script is working today, and from a code/project maintenance perspective, this seems like unnecessary risk of breaking corner cases when running the script. We already have enough CI instability, and I don't want to risk more over scripting preferences.

understood, and changed.

File: tests/scripts/validate_modified_files.sh [hunk 10]
```diff
@@ -9,6 +9,8 @@ MOD_ERR="changes found by mod.check. You may need to run make clean"
 CRD_ERR="changes found by 'make crds'. please run 'make crds' locally and update your PR"
 BUILD_ERR="changes found by make build', please commit your go.sum or other changed files"
 HELM_ERR="changes found by 'make gen-rbac'. please run 'make gen-rbac' locally and update your PR"
+DOCS_ERR="changes found by 'make docs'. please run 'make docs' locally and update your PR"
+
```
Comment (BlaineEXE) [inline tests/scripts/validate_modified_files.sh:13]: Unnecessary empty line added.

**Review by BlaineEXE — COMMENTED** [review]
Seem nice and straightforward. There's an errant empty line that is probably good to take care of, just to keep things consistent, but I don't care that much about it in the context of bash script tooling.

File: tests/scripts/validate_modified_files.sh (hunk 10 above)
Comment (obnoxxx) [reply tests/scripts/validate_modified_files.sh:13]: gone

**Review by subhamkrai — APPROVED** [review]
